text/0000-RouteInfo-Metadata.md

```diff
@@ -0,0 +1,212 @@
+- Start Date: 2018-11-02
+- RFC PR: (leave this empty)
+- Ember Issue: (leave this empty)
+
+# RouteInfo MetaData
+
+## Summary
+
+The RFC introduces the ability to associate application specific metadata with it's corresponding `RouteInfo` object. This also adds a `metadata` field to the `RouteInfo` which will be the return value of `buildRouteInfoMetadata` for it's corresponding `Route`.
+
+```js
+// app/route/profile.js
+import Route from '@ember/routing/route';
+import { inject } from '@ember/service';
+export default Route.extend({
+  user: inject('user'),
+  buildRouteInfoMetadata() {
+    return {
+      trackingKey: 'page_profile',
+      user: {
+        id: this.user.id,
+        type: this.user.type
+      }
+    }
+  }
+  // ...
+});
+```
+
+```js
+// app/services/analytics.js
+import Service, { inject } from '@ember/service';
+
+export default Service.extend({
+  router: inject('router'),
+  init() {
+    this._super(...arguments);
+    this.router.on('routeDidUpdate', (transition) => {
+      let { to, from } = transition;
+      let { metadata: fromMeta } = from;
+      let { metadata: toMeta } = to;
+      ga.sendEvent('pageView', {
+        from: fromMeta,
+        to: toMeta,
+        timestamp: Date.now(),
+      })
+    })
+  },
+  // ...
+});
+```
+
+## Motivation
+
+While the `RouteInfo` object is sufficient in providing developers metadata about the `Route` itself, it is not sufficient in layering on application specific metadata about the `Route`. This metadata could be anything from just a more domain specific name for a `Route` e.g. `profile_page` vs `profile.index`, all the way to providing contextual data when the `Route` was visited.
+
+## Detailed design
+
+### `buildRouteInfoMetadata`
+
+This optional hook is intended to be used as a way of letting the routing system know about any metadata associated with the route. It is called once any time the `Route` is entered.
```

Comment:
> It is called once any time the `Route` is entered

What exactly does `entered` mean? `routeWillChange`?

Reply:
I think we need to be super specific about when this is called by the system. It needs to be crystal clear that the model hook won't be invoked yet. I can see folks wanting to rely on data fetches in this hook.

Is it possible for this hook to be a performance problem if someone invokes a service that hasn't been instantiated yet?